[Diagnostics] Add fix-its for missing `set` and `)` after access modifier

a7medev opened the pull request:
## Summary

Added fix-its missing `set` subject and `)` for access control modifiers. The added fix-its handle the following cases:

- `private(set` inserts `)` to become `private(set)`
- `private()` inserts `set` to become `private(set)`
- `private(<unexpected>)` replaces `<unexpected>` with `set` to become `private(set)`
- `private( <declaration>` inserts `set)` to become `private(set) <declaration>`
- `private(<unexpected> <declaration>` replaces `<unexpected>` with `set)` to become `private(set) <declaration>`

a7medev committed:
[Diagnostics] Add tests for accessibility expected set fix-its

a7medev committed:
[Diagnostics] Add fix-its for missing set and ) of access modifier

a7medev commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2906,27 +2906,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+        auto SetLoc = consumeToken();
+        diag.fixItReplace(SetLoc, "set");
+        consumeToken(tok::r_paren);
+      } else if (isStartOfSwiftDecl()) {
+        diag.fixItInsertAfter(LParenLoc, "set)");
+      } else {
+        BacktrackingScope Scope(*this);
+        auto InvalidTokLoc = consumeToken();
+        
+        if (isStartOfSwiftDecl())
```
I'm using `isStartOfSwiftDecl` here to check if a declaration follows the missing `)` (optionally with an unexpected token) for the following cases:

- `private( <declaration>` inserts `set)` to be `private(set) <declaration>`
- `private(<unexpected> <declaration>` replaces `<unexpected>` with `set)` to be `private(set) <declaration>`

I'm not sure if it'd add an overhead to use it only for fix-its though. We can remove its use and not supported these two cases (or maybe replace it with a lighter check like checking for the `var` keyword).

I'd love to hear your thoughts.

a7medev replied:
@lucianopa-msft I have a quick question since you mentioned it's just a failure path, do we need the `BacktrackingScope` here or is it okay to consume the invalid tokens and not revert back at all?

a7medev replied:
Makes sense, thanks for the clarification 🙏🏼

rintaro commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2906,27 +2906,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      auto diag = diagnose(Loc, diag::attr_access_expected_set, AttrName);
+      
+      if (Tok.is(tok::l_paren)) {
+        return makeParserSuccess();
```
Special handling `private((`. Could you explain what this is for?

a7medev replied:
The previous code already checked for the current token not being `(` before consuming a `)` as it wouldn't be closing the `(` of the access modifier subject as expected in the recovery.

I thought I'd add the same check in the case of `private(( <declaration>` but thinking about it again it might not be necessary and we can still just suggest to replace the extra `(` with `set)` (we can also do the same for `private(()`, no?).

What do you think?

hamishknight replied:
It seems like the previous code was assuming if you had `private( ()` then the `()` was something unrelated to the `private(`, i.e the closing `)` matches the `(` in the `()`, not the `private(`. That seems somewhat reasonable to me, although I doubt this case happens often in practice. I think we should probably move this check to the `if (peekToken().is(tok::r_paren))` condition, that way we can still correct `private(( <declaration>`.

a7medev commented on test/attr/accessibility.swift:
```diff
@@ -69,32 +69,32 @@ package(set) // expected-note {{previous modifier specified here}}
 var unterminatedSubject = 0
 
-private(42 // expected-error{{expected 'set' as subject of 'private' modifier}} expected-error{{expected declaration}}
+private(42 // expected-error{{expected 'set' as subject of 'private' modifier}} {{9-11=set)}} expected-error{{expected declaration}}
```
There's this one already with multiple random tokens but I haven't added any assertions for the fix-its (as I didn't handle that case). What do you have in mind?

```swift
private(a bunch of random tokens)
```

a7medev replied:
I've been thinking about handling it by basically consuming tokens until we find a declaration and replace the whole range till the declaration with `set` or `set)` but there are some problems that we need to consider:

1. In the worst case, we won't have a declaration until EOF so we'll end up scanning the remaining part of the file and checking `isStartOfSwiftDecl` with each token but that's an extreme case that won't happen with a lot of remaining code anyways I think. We can also limit how far ahead we search.
2. `<access-modifier>(` may not be followed by a declaration of its own (just by mistake maybe?) and so the fix-it will try to replace irrelevant but valid tokens until the first declaration with `set` which doesn't make sense. Here's an example:
    ```swift
    private(
    // Fix-it would suggest replacing the code from here
    if isEnabled {
        startServiceA()
        startServiceB()
        performTaskC()
    }
    // till here with `set`
    var name: String = "Swift"
    ```

I'd love to hear your suggestions here.

hamishknight replied:
FWIW even if we don't handle these cases, I think it's still worth adding them as tests

a7medev commented:
@lucianopa-msft @roop I've updated the fix-its to include multiple invalid tokens (and invalid parentheses like `private((()))`), I generalized the code to not have special cases for `<access-control>()` and `<access-control>(<invalid>)` and added recovery for all cases with added fix-its.
I'd appreciate your review. 🙏🏼

hamishknight reviewed (commented):
Thanks!

a7medev commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2906,27 +2906,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+        return makeParserSuccess();
+      }
 
       // Minimal recovery: if there's a single token and then an r_paren,
       // consume them both. If there's just an r_paren, consume that.
```
The recovery still hasn't changed, do you think we should update the comment to mention the fix-its as well?

hamishknight replied:
I think it's worth mentioning the `isAtStartOfSwiftDecl` logic, but otherwise I think it's fine

a7medev replied:
Added comments explaining the fix-its logic ✅

hamishknight replied:
I would just remove this comment now that the explanations are in each of the branches

hamishknight commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2906,27 +2907,44 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      auto EndLoc = LParenEndLoc;
+
+      int OpenParens = 1;
+      while (!Tok.isAtStartOfLine() && !isStartOfSwiftDecl()) {
```
> `private((()))` turns into `private(set)(()))` which is not ideal. ❌

This seems fine to me, we have no way of knowing what the user is trying to do with `((()))` and it's very unlikely to come up in practice, I'm happy to leave it be.

> `private(some random tokens)` turns into `private(set) some random tokens)` which is not ideal. ❌

This does seem a bit more unfortunate, but again I think it's pretty unlikely to come up in practice, so I'm okay not handling it for now. If we did want to handle it, we would also need to teach swift-syntax to handle it too.

a7medev replied:
Sorry for taking so long to resolve the comments here. I've updated the PR as discussed. I've left out the case for multiple invalid tokens between the parentheses for now.

Can you please recheck? 🙏🏼

a7medev committed:
[Test] Test no fix-it for access-control set in place of unrelated code

hamishknight commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2920,27 +2920,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      } else if (isNextStartOfSwiftDecl()) {
+        // Suggest `set)` in place of an invalid token after l_paren followed by
+        // a valid declaration start.
+        // e.g. `private( var x: Int` -> `private(set) var x: Int`
+        diag.fixItReplace(Tok.getLoc(), "set)");
+      } else {
+        // Suggest `set)` after l_paren if not followed by a valid declaration
+        // e.g. `private( val x: Int` -> `private(set) val x: Int`
```
These examples seem to be the same? The first one should be something like `private(<invalid> var x: Int`, right? Also typo `val` -> `var`

a7medev replied:
`val` is intentional here, it's an invalid token. I can switch to `<invalid>` for clarity though.

hamishknight commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2920,27 +2920,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+        auto SetLoc = consumeToken();
+        diag.fixItReplace(SetLoc, "set");
+        consumeToken(tok::r_paren);
+      } else if (isNextStartOfSwiftDecl()) {
```
Given we know `Tok` is likely bogus here we should consume it

a7medev replied:
Makes sense. Done ✅

hamishknight commented on include/swift/Parse/Parser.h:
```diff
@@ -690,6 +690,13 @@ class Parser {
     return Context.LangOpts.EnableExperimentalConcurrency;
   }
 
+  /// Returns true if a a Swift declaration starts after the current token, otherwise returns false.
```
```suggestion
  /// Returns true if a Swift declaration starts after the current token, otherwise returns false.
```

hamishknight replied:
Also don't forget to run `git clang-format`, this line looks longer than the column limit

a7medev replied:
Fixed and ran `git-clang-format`, can you please recheck? 🙏🏼

hamishknight reviewed (commented):
Thank you! Couple more minor comments, but otherwise this LGTM

a7medev committed:
[IDE] Update comments for access control missing set fix-its

hamishknight commented:
@swift-ci please test

hamishknight commented:
@a7medev Looks like the test needs updating for the `consumeToken` change:

```
[2025-08-15T14:03:06.313Z] private(42 // expected-error{{expected 'set' as subject of 'private' modifier}} {{9-11=set)}} expected-error{{expected declaration}}
[2025-08-15T14:03:06.313Z]                                                                                              ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

a7medev committed:
[Test] Remove expected-error for invalid declaration after expected 'set' fix-it

a7medev commented:
@hamishknight Fixed. Can you please rerun the tests? 🙏🏼

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows
